src/types.rs

```diff
@@ -3,18 +3,47 @@
 
 use crate::ec;
 use derive_new::new;
-use std::io;
+use std::{io, os::raw::c_ulong};
 
 pub type Error = io::Error;
 pub type Result<T> = io::Result<T>;
 
 pub type MasterIndex = u32;
-pub type DomainIndex = u32;
-pub type SlaveConfigIndex = u32;
-pub type SlavePosition = u16;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub struct DomainIndex(pub(crate) usize);
+
+impl From<i32> for DomainIndex {
+    fn from(v: i32) -> Self {
+        Self(v as usize)
+    }
+}
+
+impl From<DomainIndex> for c_ulong {
+    fn from(idx: DomainIndex) -> Self {
+        idx.0 as c_ulong
+    }
+}
+
+impl From<DomainIndex> for u32 {
+    fn from(idx: DomainIndex) -> Self {
+        idx.0 as u32
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct DomainDataPlacement {
+    pub offset: usize,
```

Comment:
Should this be `DomainOffset` and `DomainSize` types? They don't seem to be used otherwise.

Reply:
Yes, I combined the two values in one struct.